actors/evm/src/interpreter/precompiles/bls12_381/g1_add.rs

```diff
@@ -0,0 +1,268 @@
+use crate::interpreter::{
+    System,
+    precompiles::{PrecompileContext, PrecompileError, PrecompileResult},
+};
+use fil_actors_runtime::runtime::Runtime;
+
+use crate::interpreter::precompiles::bls_util::{
+    G1_ADD_INPUT_LENGTH, PADDED_G1_LENGTH, encode_g1_point, extract_g1_input, is_infinity,
+    p1_from_affine, p1_to_affine,
+};
+
+use blst::{blst_p1, blst_p1_add_or_double_affine, blst_p1_affine};
+
+/// **BLS12_G1ADD Precompile**
+///
+/// Implements G1 point addition according to [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537).
+pub fn bls12_g1add<RT: Runtime>(
+    _: &mut System<RT>,
+    input: &[u8],
+    _: PrecompileContext,
+) -> PrecompileResult {
+    if input.len() != G1_ADD_INPUT_LENGTH {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    // Split the input bytes into two segments representing each G1 point.
+    let (a_bytes, b_bytes) = input.split_at(PADDED_G1_LENGTH);
+
+    // Convert the input bytes to their corresponding BLST affine representations.
+    let a_aff = extract_g1_input(a_bytes, false)?;
+    let b_aff = extract_g1_input(b_bytes, false)?;
+
+    // If either point is at infinity, return the other point as the result.
+    if is_infinity(&a_aff) {
+        return Ok(encode_g1_point(&b_aff));
+    }
+    if is_infinity(&b_aff) {
+        return Ok(encode_g1_point(&a_aff));
+    }
+
+    // Perform the addition in Jacobian coordinates, then convert back to affine.
+    let result_aff = p1_add_affine(&a_aff, &b_aff);
+    Ok(encode_g1_point(&result_aff))
+}
+
+/// Adds two G1 points given in affine form and returns the result in affine form.
+///
+/// This method is safe for point doubling since it allows `a` and `b` to be identical.
+#[inline]
+pub(super) fn p1_add_affine(a: &blst_p1_affine, b: &blst_p1_affine) -> blst_p1_affine {
+    let a_jacobian = p1_from_affine(a);
+    let sum_jacobian = p1_add_or_double(&a_jacobian, b);
+    p1_to_affine(&sum_jacobian)
+}
+
+/// Adds a G1 point in Jacobian coordinates and a G1 point in affine form.
+///
+/// Note: While this function contains an unsafe block for BLST operations,
+/// the function itself is safe because:
+/// 1. Input types (&blst_fp2) are guaranteed safe by Rust's type system
+/// 2. All possible input variants are covered by test vectors from EIP-2537
+///
+/// The unsafe block is used purely for FFI calls to the BLST library.
```

Comment:
```suggestion
/// Note: While this function contains an unsafe block for BLST operations,
/// the function itself is safe because:
/// 1. input types are all defined by blst and `repr(C)` 
/// 2. blst behavior is assumed memory safe
/// 3. The unsafe block is used purely for FFI calls to the BLST library.
```

Please use this block throughout.  I remove reference to inputs type (which was incorrect here) so that you can copy paste everywherre without worrying about fixing up.